[73] Adjust session handling

jpmcfarland opened the pull request:
See Issue #73.  Start with new Context class to store a single copy of configurations and iRODS session.

jpmcfarland commented:
This last commit contained many little changes all over the place.  Hopefully it is not too difficult to follow.

The placeholder for the session in context is an iRODS session not an iBridges session.  I ran into some kind of recursive issue with that and using that here would ignore all the convenience methods in the iBridges session anyway.

Ultimately, if we include session functionality in the singleton context, it would be far better to use the manager itself so that all the functionality is available.  This would mean that the current IrodsConnector class need not be passed as an argument everywhere.  This could unfortunately lead to potential collisions between threads.  Locking is always a possibility.

Comments are quite welcome on this last point.

jpmcfarland commented:
Trying to reconnect after `Close session` is completely broken.  I'll fix it very soon.

jpmcfarland commented:
> Trying to reconnect after `Close session` is completely broken. I'll fix it very soon.

Should be fixed.

jpmcfarland commented:
I'm not sure if all the design decisions are the most ideal, but everything works and the singleton context with connection manager is _VERY_ handy.  I wonder how well this will stand up to multiple threads.  I have some ideas on how to harden it for that if necessary.

jpmcfarland commented:
Hmm, reconnecting after `Close session` doesn't go through the full process.  It appears to be (re)using the same connection and not a new one.

wurDevTim commented:
We are already using the same irods connector from multiple threads at the same time during an up/download. So far this works without problem.

jpmcfarland commented:
> We are already using the same irods connector from multiple threads at the same time during an up/download. So far this works without problem.

That's good news!  If the iRODSSession can handle the multiple threads, the context/connector can.  The singleton instance is created in iBridges.py before any other threads have started, so that should be fine.

jpmcfarland commented:
> > Hmm, reconnecting after `Close session` doesn't go through the full process. It appears to be (re)using the same connection and not a new one.
> 
> Looking at this, the current implementation of the singleton context is incompatible with a session reset. Minimally, both the iRODS environment and connection manager will need to be effectively reset without creating new instances. I'm checking to see if this is feasible.

I have a strong feeling that the only solution will be to take the Context singleton _out_ of the global namespace so it refreshes more naturally when opening a new session.  There are multiple options.  I'm trying to find the one that works well.

jpmcfarland commented:
I believe it is finally ready.  @chStaiger, @wurDevTim, please have a look.

jpmcfarland commented:
These last changes do not affect the main functionality of the PR.  They are minor adjustments primarily related to the unit tests.

chStaiger commented:
I found some strange things.
My irods_environment_test.json contains the line `"irods_default_resource": "irodsResc"`
But I receive that on the prompt:
```
staig001@busoni iBridges-Gui % ./iBridges.py
Missing key(s) in iRODS environment: ['irods_default_resource']
Please fix and try again!
```

And the irods_connector does not detect any storage resources any longer. It seems that the context class only uses irods_environment.json and no other environment file. I could also not find out how one can pass another irods environment path to the class. @jpmcfarland Would you have a little code snippet for me to test that?

jpmcfarland commented:
> I found some strange things. My irods_environment_test.json contains the line `"irods_default_resource": "irodsResc"` But I receive that on the prompt:
> 
> ```
> staig001@busoni iBridges-Gui % ./iBridges.py
> Missing key(s) in iRODS environment: ['irods_default_resource']
> Please fix and try again!
> ```
> 
> And the irods_connector does not detect any storage resources any longer. It seems that the context class only uses irods_environment.json and no other environment file. I could also not find out how one can pass another irods environment path to the class. @jpmcfarland Would you have a little code snippet for me to test that?

That _is_ quite strange!

I have the same keyword in environment files on both my test machines.  Can you slack me your file so I can try it out?  Perhaps there is something special in it.

The default name is indeed the default iRODS name, but it should choose what the GUI gives it.  Perhaps this was caused by the first strangeness?  I will have a look.

chStaiger commented:
I am not sure whether Context should always first try to read in the default irods_environment.json. The file might not be present or might contain insufficient information for ibridges and the Context class. So you will always get logging messages like:
```
Missing key(s) in iRODS environment: ['irods_host', 'irods_user_name', 'irods_port', 'irods_zone_name', 'irods_default_resource']
```
Even though the user has not done anything.

jpmcfarland commented:
> I am not sure whether Context should always first try to read in the default irods_environment.json. The file might not be present or might contain insufficient information for ibridges and the Context class. So you will always get logging messages like:
> 
> ```
> Missing key(s) in iRODS environment: ['irods_host', 'irods_user_name', 'irods_port', 'irods_zone_name', 'irods_default_resource']
> ```
> 
> Even though the user has not done anything.

Very good point.  I'll see what I can do to adjust that.

jpmcfarland commented:
It looks like it will indeed use the one chosen connecting when actually connecting.  The default will always be `ibridges_configuration['last_ienv']` if it exists, the iRODS default filename if it doesn't.  This is chosen in `iBridges.py`.

Unfortunately, there is a place that still tries to load a non-existent file, or not finding a file, sets `ibridges_configuration = {}`.  I will see what I can do there.

chStaiger commented:
More of a general question. Do we want attributes to print information on the prompt or just return the property or None if not set?

jpmcfarland commented:
> It looks like it will indeed use the one chosen connecting when actually connecting. The default will always be `ibridges_configuration['last_ienv']` if it exists, the iRODS default filename if it doesn't. This is chosen in `iBridges.py`.
> 
> Unfortunately, there is a place that still tries to load a non-existent file, or not finding a file, sets `ibridges_configuration = {}`. I will see what I can do there.

I got rid of the concept of a default iRODS environment file name.  The user is responsible for creating it, so we don't need to test for that specific one, only find JSON files with the correct form in the default iRODS directory.  That fixes the main problem, but now something else is has come up, so I need to fix that before committing.

chStaiger commented:
I tried to have a closer look at the relationships between ContextContainer, Context, irodsConnector and session.
I find that very confusing. 
The context gets an object `irods_connector`. This one contains an object session which contains again an object `context`.
How does the session get access to the information captured in the context of the ContextContainer?

jpmcfarland commented:
The `Context` class discovers and provides the iBridges configuration and the iRODS environment.  It also provides a place to store an `IrodsConnector` instance, but leaves it to be set externally (from within iBridges.py).  The job of the `ContextContainer` class is just that: contain the useful parts of a Context instance.  It provides that functionality to any class needing it via inheritance.  Thus, any class inheriting from it gains 4 attributes (`conf`, `conn`, `context`, `ienv`).  `context` is _the_ singleton and the other three are simply shortcuts to the configuration and environment dictionaries and the connector instance.

In iBridges.py, _the_ singleton `Context` instance is created in `main()` and initialized with an `IrodsConnector` instance.  The _magic_ is that no new `Context` instance is created.  Its `__new__()` methods always reuses the same one once set the first time.  This is the case for _all_ class instances whose class inherits from it.  This includes a (recursive?) reference to the `IrodsConnector` instance within the `Resource` and `Session` class instances.  I don't believe any problems will arise from this as accessing one of their methods/properties via the `conn` property is simply a pointer to that very instance.  In any case, it is mainly the iRODS environment used in those two.

I hope this makes it clear.

jpmcfarland commented:
> @jpmcfarland Yes thank you! So whenever we use the backend (CLI, scripts, new UIs etc) we need to watch out that we create context object before the irodsConnector. I think we need to document that somewhere ;)

Once the first instance of `Context` is made, the next step is setting the `conn` property.  See the bottom of iBridges.py.  Any new startup script (e.g., iBridgesCli.py) will need to create those instances.  Also, any new _plugins_, once we have that ability, will need to inherit from the `ContextContainer` to have access to its payload.

Indeed, we will need to document our API for the plugins when we have them.  This will naturally be a part of that documentation.  We can even start that already with the relevant parts of this conversation put into a README file.

jpmcfarland commented:
I just discovered that the iRODS environment completeness check is ignored after a session reset.  I'll need to fix that.

jpmcfarland commented:
> I just discovered that the iRODS environment completeness check is ignored after a session reset. I'll need to fix that.

Fixed!

chStaiger commented:
I do have some worrying observations when looking at the current backend. I followed the code in the iBridges.py

When I init the context and like we do there:
```
In [1]: from irodsConnector.manager import IrodsConnector
In [2]: from utils.context import Context
In [3]: context = Context()
In [4]: context.irods_connector = IrodsConnector()
```
I can get an infinite procession of `session.conn`, which I am not sure is very handy
In [5]: context.irods_connector.session.conn.session.conn
Out[5]: <irodsConnector.manager.IrodsConnector at 0x107067850>
```

Another thing that worries me is that we can have empty properties at different layers of the objects that we create that make testing, whether some settings are parsed correctly very difficult. E.g. getting the host from a session in an iRODS connector can lead to an unexpected exception, which we would have to check for every time we use the irods connector

```
In [6]: context.irods_connector.session.host
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 context.irods_connector.session.host

File ~/git-repos/iBridges-Gui/irodsConnector/session.py:75, in Session.host(self)
     65 @property
     66 def host(self) -> str:
     67     """Retrieve hostname of the iRODS server
     68
     69     Returns
   (...)
     73
     74     """
---> 75     return self._session.host

AttributeError: 'NoneType' object has no attribute 'host'
```

I think we need to brainstorm about which classes should do some checks on the session object and also on the context class members.

jpmcfarland commented:
> I do have some worrying observations when looking at the current backend. I followed the code in the iBridges.py
> 
> When I init the context and like we do there:
> 
> ```
> In [1]: from irodsConnector.manager import IrodsConnector
> In [2]: from utils.context import Context
> In [3]: context = Context()
> In [4]: context.irods_connector = IrodsConnector()
> ```
> 
> I can get an infinite procession of `session.conn`, which I am not sure is very handy

This is easily fixed by manually adding the context and select convenience properties to IrodsConnector, Session, and Resource such that they do not inherit from ContextContainer.  This is only done now to avoid code duplication with the unfortunate side-effect you have pointed out.

jpmcfarland commented:
> Another thing that worries me is that we can have empty properties at different layers of the objects that we create that make testing, whether some settings are parsed correctly very difficult. E.g. getting the host from a session in an iRODS connector can lead to an unexpected exception, which we would have to check for every time we use the irods connector
> 
> ```
> In [6]: context.irods_connector.session.host
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> Input In [6], in <cell line: 1>()
> ----> 1 context.irods_connector.session.host
> 
> File ~/git-repos/iBridges-Gui/irodsConnector/session.py:75, in Session.host(self)
>      65 @property
>      66 def host(self) -> str:
>      67     """Retrieve hostname of the iRODS server
>      68
>      69     Returns
>    (...)
>      73
>      74     """
> ---> 75     return self._session.host
> 
> AttributeError: 'NoneType' object has no attribute 'host'
> ```

This is caused simply by me being sloppy.  A check for the population of the shadow attribute should be included in all these properties.  I only have added these checks where I had errors.  I can go through and update this throughout.

Do you think that will solve it or do you think there is a more fundamental issue?

jpmcfarland commented:
I have updated the `irodsConnector/*py` properties.  They were generally fine, but needed a little attention.

jpmcfarland commented:
@chStaiger have your current concerns been addressed by the recent commits?

chStaiger commented:
> @chStaiger have your current concerns been addressed by the recent commits?

I still have the problem that the GUI does not detect the resources correctly:
```
    -=WARNING=-        -=WARNING=-        -=WARNING=-        -=WARNING=-
The default resource (irodsResc) not found in available resources!
Check "irods_default_resource" and "force_unknown_free_space" settings.
    -=WARNING=-        -=WARNING=-        -=WARNING=-        -=WARNING=-
```
The status tab also does not show any resources.

Although, when I add in line 153 in iBridges.py
```
print(self.conn.resources)
```
I get the full list of resources.

chStaiger commented:
> > @chStaiger have your current concerns been addressed by the recent commits?
> 
> I still have the problem that the GUI does not detect the resources correctly:
> 
> ```
>     -=WARNING=-        -=WARNING=-        -=WARNING=-        -=WARNING=-
> The default resource (irodsResc) not found in available resources!
> Check "irods_default_resource" and "force_unknown_free_space" settings.
>     -=WARNING=-        -=WARNING=-        -=WARNING=-        -=WARNING=-
> ```
> 
> The status tab also does not show any resources.
> 
> Although, when I add in line 153 in iBridges.py
> 
> ```
> print(self.conn.resources)
> ```
> 
> I get the full list of resources.

Can it be that, the resource.py is ignoring all resources when there is no `free_space` set? 
```
105                 if metadata['parent'] is None and metadata['free_space'] > 0:
106                     resc_names.append(name)
```

jpmcfarland commented:
@wurDevTim Please check if your iRODS session creation has been improved after the last commit.  I have removed extraneous keywords so only those from the iRODS environment file should be passed to the iRODSSession constructor.

jpmcfarland commented:
> Can it be that, the resource.py is ignoring all resources when there is no `free_space` set?
> 
> ```
> 105                 if metadata['parent'] is None and metadata['free_space'] > 0:
> 106                     resc_names.append(name)
> ```

Yes, but this check is informational only.  I can have the warning be seen only when the `force_unknown_free_space` flag is False to avoid confusion.  Also, this code is actually at lines 119 and 120.  Please check yours.

jpmcfarland commented:
Sorry for the disruption, but the only way I found to _properly_ reconcile the logic for resource checking was to split the `force_unknown_free_space` option into `check_free_space` and `force transfers` options.  I'm sure some additional tweaking will be necessary in the functionality.

Please feel free to comment.

wurDevTim commented:
> @wurDevTim Please check if your iRODS session creation has been improved after the last commit. I have removed extraneous keywords so only those from the iRODS environment file should be passed to the iRODSSession constructor.

(https://github.com/chStaiger/iBridges-Gui/blob/731348b3e9958e72fd3ea93f67d947edf89082c6/irodsConnector/session.py#L286)
misses the _ = options.pop('irods_env_file')

After this change we can connect to the servers at SURF SARA again. 
But the irodsbrowsers remains empty, when I look in the data operaions call from 
https://github.com/chStaiger/iBridges-Gui/blob/731348b3e9958e72fd3ea93f67d947edf89082c6/gui/IrodsBrowser.py#L307
I see that you are now passing an utils.path.IrodsPath, the path exists on the server and is correct.   

![image](https://user-images.githubusercontent.com/81219429/232738564-9d123178-454e-4f40-a8ba-7d20d985ab64.png)

jpmcfarland commented:
> > @wurDevTim Please check if your iRODS session creation has been improved after the last commit. I have removed extraneous keywords so only those from the iRODS environment file should be passed to the iRODSSession constructor.
> 
> (
> 
> https://github.com/chStaiger/iBridges-Gui/blob/731348b3e9958e72fd3ea93f67d947edf89082c6/irodsConnector/session.py#L286
> 
> )
> misses the _ = options.pop('irods_env_file')
> After this change we can connect to the servers at SURF SARA again.

Will moving the `pop()` outside of the conditional do the job?  For instance:

```python
        irods_env_file = options.pop('irods_env_file')
        if 'password' not in options:
            try:
                print('AUTH FILE SESSION')
                session = irods.session.iRODSSession(
                    irods_env_file=irods_env_file)
```

wurDevTim commented:
@jpmcfarland, that works. It returns a valid session.
Do you also have an empty irods browser?

jpmcfarland commented:
> But the irodsbrowsers remains empty, when I look in the data operaions call from
> 
> https://github.com/chStaiger/iBridges-Gui/blob/731348b3e9958e72fd3ea93f67d947edf89082c6/gui/IrodsBrowser.py#L307
> 
> 
> I see that you are now passing an utils.path.IrodsPath, the path exists on the server and is correct.
> ![image](https://user-images.githubusercontent.com/81219429/232738564-9d123178-454e-4f40-a8ba-7d20d985ab64.png)

All `*Path` objects are `str` objects and `IrodsPath`s are additionally `iRODSPath` objects (which are also `str` objects).  They _MUST_ be compatible!

I will do some testing to see if I can figure it out.  As a work-around, just cast back to `str` for now.

wurDevTim commented:
> > But the irodsbrowsers remains empty, when I look in the data operaions call from
> > https://github.com/chStaiger/iBridges-Gui/blob/731348b3e9958e72fd3ea93f67d947edf89082c6/gui/IrodsBrowser.py#L307
> > 
> > I see that you are now passing an utils.path.IrodsPath, the path exists on the server and is correct.
> > ![image](https://user-images.githubusercontent.com/81219429/232738564-9d123178-454e-4f40-a8ba-7d20d985ab64.png)
> 
> All `*Path` objects are `str` objects and `IrodsPath`s are additionally `iRODSPath` objects (which are also `str` objects). They _MUST_ be compatible!
> 
> I will do some testing to see if I can figure it out. As a work-around, just cast back to `str` for now.

Too help figure out where to look: the issue is new in this branch. The irods browser is populated correctly on the development branch.

wurDevTim commented:
@jpmcfarland, after some debugging we discovered that the code works on Ubuntu, not on windows. The cause is most likely the different wrapping of the pathlib library. We noticed that you removed the `is_posix`, and the different implementations for windows and Ubuntu/IOS.

jpmcfarland commented:
> @jpmcfarland, that works. It returns a valid session.
> Do you also have an empty irods browser?

I need to check.  I just got back from a minor auto pech.

jpmcfarland commented:
> @jpmcfarland, after some debugging we discovered that the code works on Ubuntu, not on windows. The cause is most likely the different wrapping of the pathlib library. We noticed that you removed the `is_posix`, and the different implementations for windows and Ubuntu/IOS.

While going back through the path module, I realized that the POSIX check was already done by the `pathlib` library itself and I was simply repeating it.  `PurePath`s and `LocalPath`s should be compatible with the platform by design.

wurDevTim commented:
> ```python
> str(utils.path.IrodsPath("/npec/home/mobile")) == utils.path.IrodsPath("/npec/home/mobile")
> ```
returns False, 
![image](https://user-images.githubusercontent.com/81219429/232977034-b60f02a1-fbb1-4247-8635-c8b351c35f4b.png)

I have some plants in the greenhouse which keep growing so I first have to fix some things there. When that is done I will take a look at pathlib. I am also curious about what pathlib is transmitting to the irods server, I have an idea of how to figure that out.

jpmcfarland commented:
> > ```python
> > str(utils.path.IrodsPath("/npec/home/mobile")) == utils.path.IrodsPath("/npec/home/mobile")
> > ```
> 
> returns False, ![image](https://user-images.githubusercontent.com/81219429/232977034-b60f02a1-fbb1-4247-8635-c8b351c35f4b.png)
> 

I see enough in your screen capture to convince me it is a library difference issue.  I'll see if I can find a fix or at least a proper workaround for Windows.

jpmcfarland commented:
> > ```python
> > str(utils.path.IrodsPath("/npec/home/mobile")) == utils.path.IrodsPath("/npec/home/mobile")
> > ```
> 
> returns False, ![image](https://user-images.githubusercontent.com/81219429/232977034-b60f02a1-fbb1-4247-8635-c8b351c35f4b.png)

It is _effectively_ a PRC bug under Windows, but only when using an `IrodsPath` object, thus _our_ bug:

```python
print(str(utils.path.IrodsPath('/a/b/c')))
/a/b/c
print(repr(str(utils.path.IrodsPath('/a/b/c'))))
'/a/b/c'
print(irods.path.iRODSPath(utils.path.IrodsPath('/a/b/c')))
/\a\b\c
print(repr(irods.path.iRODSPath(utils.path.IrodsPath('/a/b/c'))))
'/\\a\\b\\c'
```

In the `get()` method of a `CollectionManager` the input path is "normalized".  It is the same process the second two commands above illustrate.  Instead of the query receiving `[Collection.name == '/a/b/c']`, it gets `[Collection.name == '/\\a\\b\\c']`, causing the missing collections.

The workaround is to cast `IrodsPath`s to `str`s, but the whole point of the `IrodsPath` object is to prevent this need in the first place!  It effectively turns a `pathlib.Path` object into a `str` object with extra functionality.

Let me see if I can update `IrodsPath` to fix this.  Not likely, but worth a try.

jpmcfarland commented:
Here is the problem: `p = '/'.join(paths).split('/')` (line 63 of irods/path/__init__.py).  When encountering an `IrodsPath` of ours, its string part is converted from a `pathlib.PurePosixPath` to a `pathlib.PureWindowsPath` during the `join()` and the split sees no `/`s so the expected split does not occur.

```python
paths = (utils.path.IrodsPath("/a/b/c"),)
p = '/'.join(paths).split('/')
print(p)
['', '\\a\\b\\c']
```

Removing the strict POSIX treatment of the `*Paths` is the ultimate cause.  Restoring this treatment fixes the problem:

```python
paths = (utils.path.IrodsPath("/a/b/c"),)
p = '/'.join(paths).split('/')
print(p)
['', 'a', 'b', 'c']
```

d-w-moore commented:
@jpmcfarland - is `util.path.IrodsPath` a passthru for (or perhaps a copy of)  `irods.path.iRODSPath`?  Forgive my intrusion here, I was just curious ... the present incarnation of iRODSPath in PRC has reverted all POSIX (or indeed OS dependent) usages and relies strictly on string handling to do its work, fwiw : )  I think it was for similar reasons that I had to do this reversion.....

jpmcfarland commented:
> @jpmcfarland - is `util.path.IrodsPath` an alias for (or perhaps a copy of) `irods.path.iRODSPath`? Forgive my intrusion here, I was just curious ... the present incarnation of `iRODSPath` in the Python iRODS Client has reverted all POSIX (or indeed OS dependent) usages and relies strictly on string handling to do its work, fwiw : ) I think it was for similar reasons that I had to do this reversion.....

Hi @d-w-moore, not a problem!

The similarity is in name only.  I have created a `PurePath` class that is a `str` and uses a `pathlib.PurePath` shadow property to mirror `pathlib` functionality but be seen as a `str` instance.  An `IrodsPath` and `LocalPath` derive from it.  Additionally, `IrodsPath` derives from `irods.path.iRODSPath` as well for _its_ custom path normalization.  `pathlib` has its own path normalization, but it is not quite the same.

There are checks in the PRC that expect a `str`, but a `pathlib.PurePath` does not derive from `str`, so in order to have pathlib functionality without having to cast it as a `str` before passing to a PRC function, an approach like this was necessary.  Unfortunately, it took alot longer to sort out the little details like the one I reverted back to above, but it has been very educational.

d-w-moore commented:
@jpmcfarland - well, that's certainly some black belt Python right there : )  .  I'd heard of `pathlib` too, but at the time I wrote the `iRODSPath` class, I think I hadn't the depth of understanding to see what advantages it might offer.

d-w-moore commented:
If it helps, I do know that `__eq__` gets called from the most-derived class, so if  you have `class PathObj(str)`, then `"a"==PathObj("a")`  and `PathObj("a")=="a"` both evaluate to `True`, and both call `PathObj.__eq__` ... but you probably already knew that : )

jpmcfarland commented:
Indeed!

I certainly stick to the inherited methods when they are doing the correct thing, as I believe they are here.  Unfortunately, the problem above is not with the comparison, but that the path string given to `str.__new__(path.__str__())` is from the wrong type of `pathlib` path.  It is _inherently incorrect_.  Why it is fine on the initial instance but not on the new instance (created by `join()`, I believe), I don't fully understand (Python Voodoo, perhaps?).  I just make sure it is always the correct type now.

chStaiger commented:
I did a fresh pull this afternoon and found that the writing of the password does not work. Testing shows that
this if statement always evaluates to False:

(session.py)
```
227             try:
228                 # Check for good authentication and cache PAM password
229                 if 'password' in options:
230                     _ = self._session.server_version
231                     self._write_pam_password()
232             except (irods.exception.CAT_INVALID_AUTHENTICATION, KeyError) as er    ror:
233                 raise error
```
 That is caused by `_get_irods_session(options)`

```
284         else:
285             password = options.pop('password')
```

Is it handy to move the password writing outside if the `if`?

chStaiger commented on README.md:
```diff
@@ -45,41 +45,43 @@ python3.10 -m pip install -r requirements.txt
-    "force_unknown_free_space": false 
+    "check_free_space": true,
+    "davrods_server": "https://server.fqdn.nl",
+    "force_transfers: false",
```
```suggestion
    "force_transfers": false",
```

jpmcfarland commented:
@chStaiger In this _last_ commit, I removed the extra session validation because it was redundant.  `_get_irods_session()` _always_ returns a validated session or `None`.

jpmcfarland commented:
@wurDevTim Just wondering how close you are to finalizing your testing.  The longer I look at it the more little non-critical, but ultimately necessary tweaks I find.

wurDevTim commented:
@jpmcfarland I saw the commits and the nice discussion with Moore, but because it's a big PR I need to find some time to really look at it. In general I am noticing that we don't test enough, which results in broken functionality on the development branch. This week I also found some related to the session, on monday I blocked a few hours to test your branch again on a windows machine. 

Unfortunately that's the earliest moment for me, there is too much going on at my end at the moment.

jpmcfarland commented:
No worries!  I just wanted to check.  I have other efforts to keep me more than busy enough at the moment.  Thanks for the update!

wurDevTim commented:
The pyinstaller script was broken in previous changes, I fixed them in order to test your changes in the paths. That looks good now. I still have an issue with the PAM password caching but Christine also tested it on her servers and there it works, so that's probably something in our config. 

In the logging its a little confusing that you always try to start a 'PAM' session first, even if 'PAM' is not in the irods keywords or supported by the server.

It's also not clear to me why you removed the underscores from the private variables? I thought that is how we are supposed to do it in python?

jpmcfarland commented:
> In the logging its a little confusing that you always try to start a 'PAM' session first, even if 'PAM' is not in the irods keywords or supported by the server.

The PAM check has always been first with the fallback.  I integrated it pretty much as is into the connector.  I can certainly update the logic so that it uses the `irods_authentication_scheme` keyword to dictate the session creation method, but that should definitely be a separate issue/pull request.

> It's also not clear to me why you removed the underscores from the private variables? I thought that is how we are supposed to do it in python?

Part was for consistency with other modules, part was to distinguish between normal local variables and shadow variables used with properties.  For instance:

```python
class A:
    _name = None

    @property
    def name(self):
        if self._name is None:
            # do something
        return self._name
```

If there is a variable that should not be accessed from outside a module that easily could, indicating it private with a leading underscore is the correct method.  I believe they should generally be used sparingly, however.  I will look into some documentation of best practices for using private naming.

jpmcfarland commented:
> I still have an issue with the PAM password caching but Christine also tested it on her servers and there it works, so that's probably something in our config.

Weird!  I have no problems on all three different OSes.  The ~/.irods/.irodsA file is cached and used properly on them all.  I have just updated some code in that area.  Was this issue seeen before or after that change?

jpmcfarland commented:
> > In the logging its a little confusing that you always try to start a 'PAM' session first, even if 'PAM' is not in the irods keywords or supported by the server.
> 
> The PAM check has always been first with the fallback. I integrated it pretty much as is into the connector. I can certainly update the logic so that it uses the `irods_authentication_scheme` keyword to dictate the session creation method, but that should definitely be a separate issue/pull request.

Looking into this, I see that if no password is given, the auth file is used for authentication.  If one is given, the contents of the `irods_environment.json` file are sent as keyword arguments to the session constructor with no priority for any authentication scheme.  A `PlainTextPAMPasswordError` is excepted in the case this fails.  No other specific `Exception`s are captured and a set of SSL settings are added to the keywords when the one named `Exception` is captured, and the session creation is attempted for the last time.

I see no code indicating a PAM session attempt.  What errors are you getting and at what stage?

wurDevTim commented:
> > I still have an issue with the PAM password caching but Christine also tested it on her servers and there it works, so that's probably something in our config.
> 
> Weird! I have no problems on all three different OSes. The ~/.irods/.irodsA file is cached and used properly on them all. I have just updated some code in that area. Was this issue seeen before or after that change?

Its very strange, when you just added the functionality, right after 'the big merge', it worked. We are still using version 1.1.6 of the python irods-client. But we did upgrade our servers from 4.2.11 to 4.3.0. And now the pam_password from a valid session is wrong:
```    
def _write_pam_password(self):
        """Store the returned PAM/LDAP password in the iRODS
        authentication file in obfuscated form.

        """
        connection = self._session.pool.get_connection()
        pam_passwords = self._session.pam_pw_negotiated
```
f24c1c15c1d16e7e937* I changed a few of the numbers to ensure it cannot be decrypted. but left the letters, the decryption of the python irod-client looks at character 6, 


The suprising thing is that the icommands password caching does work on our servers (if we change the "irods_authentication_scheme" from "pam" to  "pam_password". 
Christine told me she is sill using irods v 4.2.11 and for her the python password caching works. which makes me think it's an `irods problem`, not something in this PR. 

P.s. I saw the obfuscation code was updated in python irods-client v1.1.7, but its still responding the same.

jpmcfarland commented:
> Christine told me she is sill using irods v 4.2.11 and for her the python password caching works. which makes me think it's an `irods problem`, not something in this PR.

Indeed.  We should somehow annotate full compatibility only up to iRODS 4.2.11.

d-w-moore commented:
> The suprising thing is that the icommands password caching does work on our servers (if we change the "irods_authentication_scheme" from "pam" to "pam_password".
> Christine told me she is sill using irods v 4.2.11 and for her the python password caching works. which makes me think it's an irods problem, not something in this PR.

Not to distract from this thread, but I wanted to say that, if you find investigation is needed into the handling of PAM logins in PRC especially regarding iRODS 4.2.x, let us know - probably best via an email to `support@irods.org`. Note v1.1.8 of the PRC is due to be released soon, with as yet no PAM-related changes planned, but I'm of course interested in knowing specific and compelling evidence of any problems in that area!

d-w-moore commented:
I understand! thanks @wurDevTim

wurDevTim reviewed (approved):
The basics functionality works on windows.

jpmcfarland commented:
It is bizarre that the conversations only showed up here _after_ the approval.  I have updated them nonetheless.
